File: algovisual/frontend/src/merge.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation

# Merge Sort Algorithm
def merge_sort(arr, left, right, bars, ax):
    if left < right:
        mid = (left + right) // 2
        merge_sort(arr, left, mid, bars, ax)
        merge_sort(arr, mid + 1, right, bars, ax)
        merge(arr, left, mid, right, bars, ax)

# Merging two halves of the array
def merge(arr, left, mid, right, bars, ax):
    i = left
    j = mid + 1
    k = left
    temp_arr = arr.copy()
    
    while i <= mid and j <= right:
        if arr[i] <= arr[j]:
            temp_arr[k] = arr[i]
            i += 1
        else:
            temp_arr[k] = arr[j]
            j += 1
        k += 1

    while i <= mid:
        temp_arr[k] = arr[i]
        i += 1
        k += 1

    while j <= right:
        temp_arr[k] = arr[j]
        j += 1
        k += 1

    for i in range(left, right + 1):
        arr[i] = temp_arr[i]
    
    # Update the bars to reflect the current state of the array
    for rect, h in zip(bars, arr):
        rect.set_height(h)
    
    ax.set_title(f"Merge Sort: {arr}")
    plt.draw()

# Set up the figure and axis for the plot
fig, ax = plt.subplots(figsize=(8, 6))
ax.set_xlim(-0.5, 4.5)
ax.set_ylim(0, 10)

# Generate 5 random values between 1 and 10
arr = np.random.randint(1, 11, size=5)
bars = ax.bar(range(len(arr)), arr)

# Create the animation
def animate(i):
    merge_sort(arr, 0, len(arr) - 1, bars, ax)

ani = FuncAnimation(fig, animate, frames=1, repeat=False, interval=2000)

# Display the plot
plt.show()

File: algovisual/frontend/src/merge.js
let values = [];
let states = [];
let done = false;

function setup() {
  createCanvas(600, 400); 
  values = new Array(5).fill(0).map(() => random(height));
  states = new Array(values.length).fill(-1);
  
  mergeSort(values, 0, values.length - 1);
}

function draw() {
  background(255);
  for (let i = 0; i < values.length; i++) {
    if (states[i] === -1) fill(0);
    else if (states[i] === 0) fill(0, 0, 255); 
    else if (states[i] === 1) fill(0, 255, 0); 
    noStroke(); 
    rect(i * (width / values.length), height - values[i], width / values.length - 4, values[i]);
  }

  if (done) noLoop();
}

async function mergeSort(arr, start, end) {
  if (start >= end) {
    if (start >= 0) states[start] = 1; 
    return;
  }
  const mid = Math.floor((start + end) / 2);
  await mergeSort(arr, start, mid);
  await mergeSort(arr, mid + 1, end);
  await merge(arr, start, mid, end);
}

async function merge(arr, start, mid, end) {
  let left = arr.slice(start, mid + 1);
  let right = arr.slice(mid + 1, end + 1);
  
  let i = 0, j = 0, k = start;

  states[start] = 0; 

  while (i < left.length && j < right.length) {
    await sleep(1000); 
    if (left[i] < right[j]) {
      arr[k++] = left[i++];
    } else {
      arr[k++] = right[j++];
    }
  }

  while (i < left.length) {
    await sleep(1000);
    arr[k++] = left[i++];
  }

  while (j < right.length) {
    await sleep(1000);
    arr[k++] = right[j++];
  }

  if (start >= 0) {
    for (let l = start; l <= end; l++) states[l] = 1; 
  }

  if (start === 0 && end === arr.length - 1) done = true; 
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}